### PR TITLE
GLTFLoader: Preserve unknown extension data on nodes and materials.

### PR DESCRIPTION
**Pull request by donmccurdy**
Fixes https://github.com/mrdoob/three.js/issues/13703 and https://github.com/mrdoob/three.js/issues/11682. Not a complete solution for all user-defined extensions — may need access to data in buffers or textures — but if that is requested later I think we can just return a `parser` reference and then custom extension implementations can do the rest pretty easily:

```js
loader.load('foo.gltf', (gltf) => {
  var scene = gltf.scene;
  var mesh = scene.children[ 3 ];
  var fooExtension = mesh.userData.gltfExtensions.EXT_foo;
  gltf.parser.getDependency( 'bufferView', fooExtension.bufferView )
    .then( fooBuffer => { ... } );
});
```

This will also avoid loading those textures and buffers until they're requested by the userland code, so it seems pretty decent as an API.

/cc @takahirox

### CODE REVIEW

File: examples/js/loaders/GLTFLoader.js [hunk 1]
```diff
@@ -123,27 +123,32 @@ THREE.GLTFLoader = ( function () {
...
-					extensions[ EXTENSIONS.KHR_MATERIALS_PBR_SPECULAR_GLOSSINESS ] = new GLTFMaterialsPbrSpecularGlossinessExtension();
-
-				}
+						case EXTENSIONS.KHR_DRACO_MESH_COMPRESSION:
+							extensions[ extensionName ] = new GLTFDracoMeshCompressionExtension( this.dracoLoader );
+							break;
 
-				if ( json.extensionsUsed.indexOf( EXTENSIONS.KHR_DRACO_MESH_COMPRESSION ) >= 0 ) {
+						default:
+							console.warn( 'THREE.GLTFLoader: Unknown extension "' + extensionName + '".' );
```
Comment (stevesan) [inline examples/js/loaders/GLTFLoader.js]: Not big deal but: Should this be a warning? If the intent is to let users implement their own extensions, wouldn't we expect this to happen regularly? Perhaps another approach is to let users pass in a list of extensions to expect, so 'extensionsUsed' can be meaningful warnings and 'extensionsRequired' can be meaningful errors.

File: examples/js/loaders/GLTFLoader.js [hunk 2]
```diff
@@ -1128,6 +1133,23 @@ THREE.GLTFLoader = ( function () {
 
 	}
 
+	function addExtensionUserData( extensionsUsed, object, objectDef ) {
```
Comment (stevesan) [inline examples/js/loaders/GLTFLoader.js]: better name may be: "addUnknownExtensionsToUserData"

File: examples/js/loaders/GLTFLoader.js [hunk 3]
```diff
@@ -2654,6 +2678,8 @@ THREE.GLTFLoader = ( function () {
 
 			if ( nodeDef.extras ) node.userData = nodeDef.extras;
```
Comment (stevesan) [inline examples/js/loaders/GLTFLoader.js]: It's true that any GLTF object can have extensions right? Material is most immediately useful for us, but just 'scene' would also be useful.

**Review by stevesan — COMMENTED** [review]
Thanks a ton!

File: examples/js/loaders/GLTFLoader.js [hunk 4]
```diff
@@ -123,27 +123,38 @@ THREE.GLTFLoader = ( function () {
...
-				if ( json.extensionsUsed.indexOf( EXTENSIONS.KHR_MATERIALS_PBR_SPECULAR_GLOSSINESS ) >= 0 ) {
+						case EXTENSIONS.KHR_DRACO_MESH_COMPRESSION:
+							extensions[ extensionName ] = new GLTFDracoMeshCompressionExtension( this.dracoLoader );
+							break;
 
-					extensions[ EXTENSIONS.KHR_MATERIALS_PBR_SPECULAR_GLOSSINESS ] = new GLTFMaterialsPbrSpecularGlossinessExtension();
+						default:
 
-				}
+							if ( extensionsRequired.indexOf( extensionName ) >= 0 ) {
```
Comment (takahirox) [inline examples/js/loaders/GLTFLoader.js]: Why is this `if` necessary?

**Comment by takahirox** [comment]
Looks good first workaround. And we should update glTF loader documentation for this user defined extensions.

File: examples/js/loaders/GLTFLoader.js [hunk 5]
```diff
@@ -1128,6 +1139,23 @@ THREE.GLTFLoader = ( function () {
 
 	}
 
+	function addUnknownExtensionsToUserData( object, objectDef ) {
+
+		// Add unknown glTF extensions to an object's userData.
+
+		for ( var name in objectDef.extensions ) {
+
+			if ( EXTENSIONS[ name ] === undefined ) {
```
Comment (takahirox) [inline examples/js/loaders/GLTFLoader.js]: keys in `EXTENSIONS` are `GLTFLoader`-defined keys, not actual extensions names.

	var EXTENSIONS = {
		KHR_BINARY_GLTF: 'KHR_binary_glTF',
		KHR_DRACO_MESH_COMPRESSION: 'KHR_draco_mesh_compression',
		KHR_LIGHTS: 'KHR_lights',
		KHR_MATERIALS_PBR_SPECULAR_GLOSSINESS: 'KHR_materials_pbrSpecularGlossiness',
		KHR_MATERIALS_UNLIT: 'KHR_materials_unlit'
	};

File: examples/js/loaders/GLTFLoader.js (hunk 3 above)
Comment (stevesan) [reply examples/js/loaders/GLTFLoader.js]: Awesome! 'mesh' may also be useful, but not immediate for me.

**Comment by mrdoob** [comment]
@takahirox looks good?

File: examples/js/loaders/GLTFLoader.js (hunk 4 above)
Comment (takahirox) [reply examples/js/loaders/GLTFLoader.js]: Personally I prefer (a) because users can know all extensions ignored but we can discuss later.

**Comment by takahirox** [comment]
Why do you target only `scene`, `mesh`, and `node` tho other glTF objects also can have `.extensions` field, for example `primitive(geometry)` and `textures`? Because Three.js objects corresponding to them don't have `.userData` property?

**Comment by donmccurdy** [comment]
I haven't made the decision on an object-by-object basis here, I'm just adding the three (`scene`, `node`, `material`) that seemed most useful. I omit meshes and primitives because they don't have a stable mapping to three.js objects — I don't want to put something in place that is likely to be unreliable, when no one is yet asking for it. We can consider adding more objects if there's a good reason.

**Comment by takahirox** [comment]
Hm, as of me `texture` (and maybe `image`) would be also useful. But yeah we can consider later when someone actually requests. And probably we need to think returning `parser` option if we consider `texture` extension.

**Comment by takahirox** [comment]
I want comment in the code or document that we support only three(`scene`, `node`, `material`) extensions.

**Comment by takahirox** [comment]
Can we expose top level `json.extensions`?

VRM, VR 3D avatar format based on glTF 2.0, is released today.

https://dwango.github.io/vrm/
https://dwango.github.io/vrm/vrm_spec/

VRM extension is defined in top level `json.extensions`. I'm gonna make `VRMLoader` using `GLTFLoader`.

Where should we store `json.extensions`? `scene.userData`?

**Comment by donmccurdy** [comment]
Sure, I'll add `.extensions` to the `onLoad` callback's object? Already have extensions from the scene object on `scene.userData`.

**Comment by takahirox** [comment]
Yeah, let's do that.

**Comment by takahirox** [comment]
I realized `parser.getDependency()` is necessary for VRM.... Can we go with returning `parser` now?

**Comment by donmccurdy** [comment]
Ok, updated. 👍

File: docs/examples/loaders/GLTFLoader.html [hunk 6]
```diff
@@ -84,6 +84,28 @@ <h2>Browser compatibility</h2>
...
+
+  var mesh = scene.children[ 3 ];
+
+  var fooExtension = mesh.userData.gltfExtensions.EXT_foo;
+
+  gltf.parser.getDependency( 'bufferView', fooExtension.bufferView )
+    .then( function ( fooBuffer ) { ... } );
+
+});
+		</code>
```
Comment (takahirox) [inline docs/examples/loaders/GLTFLoader.html]: Why you don't add indentations?

**Review by takahirox — APPROVED** [review]

**Comment by donmccurdy** [comment]
Quick bump if this looks ok? 🙂

**Comment by mrdoob** [comment]
Looks good to me!

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**